### PR TITLE
Homework 2 Complete

### PR DESCRIPTION
**Pull request by Shadragon99**
I have completed all the work and believe it should be satisfactory

### CODE REVIEW

File: src/main/java/Problem1.java [hunk 1]
```diff
@@ -1,10 +1,62 @@
...
+
+                    center ++;
+
+            }
+
+            else if (nums[center] == 2) {
+
+                temp = nums[center];// temp allows for the center and end positions to switch places
+                nums[center] = nums[end];
+                nums[end] = temp;
```
Comment (pdgetrf) [inline src/main/java/Problem1.java:52]: duplicated logic for swapping numbers.

refactor this input a private swap function

File: src/main/java/Problem2.java [hunk 2]
```diff
@@ -1,12 +1,40 @@
...
+                if (nums[i] == maxnum){
+                    break;
+                }
+                for (int j = i + 1; j < nums.length-1; j++) {
+                    nums[j] = nums[j + 1];
+
+                }
+            }
+        }
+        for (int l = 0; l < nums.length; l++){
```
Comment (pdgetrf) [inline src/main/java/Problem2.java:30]: can the length be inferred from the loop above?

File: src/main/java/Problem3/Book.java [hunk 3]
```diff
@@ -16,10 +16,26 @@ public Book(String title, String author) {
...
+    public boolean equals(Object obj) { //had help from https://www.geeksforgeeks.org/overriding-equals-method-in-java/
         // homework
+        if (obj == this){
+            return true;
+        }
+        if (!(obj instanceof Book)){
+            return false;
+        }
+        Book b1 = (Book) obj;
+        return (this.id == b1.id && this.author == b1.author && this.title == b1.title);
```
Comment (pdgetrf) [inline src/main/java/Problem3/Book.java:39]: wrong logic 

```
A movie equals another movie only when their id are the same. Same rule for books being equal.
```

File: src/main/java/Problem3/Book.java [hunk 4]
```diff
@@ -16,10 +16,26 @@ public Book(String title, String author) {
     // copy constructor
     public Book(Book anotherBook) {
         // homework
+        this.title = anotherBook.title;
+        this.author = anotherBook.author;
+        this.id = anotherBook.id;
     }
 
+    public Book() {
```
Comment (pdgetrf) [inline src/main/java/Problem3/Book.java:24]: is this used anywhere?

File: src/main/java/Problem3/BookFiction.java [hunk 5]
```diff
@@ -1,20 +1,32 @@
...
     private String genres;
 
     public BookFiction(String title, String author, String genres) {
+        super();
         // homework
         // tip: use the 'super' keyword
+        super.title = title;
+        super.author = author;
+        this.genres = genres;
+        super.id = UUID.randomUUID();
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookFiction.java:18]: wrong way of using logic and redundant id generation

File: src/main/java/Problem3/BookFiction.java [hunk 6]
```diff
@@ -1,20 +1,32 @@
...
     }
 
     // copy constructor
     public BookFiction(BookFiction anotherBook) {
+        super();
         // homework
         // tip: use the 'super' keyword
+        super.title = anotherBook.title;
+        super.author = anotherBook.author;
+        super.id = anotherBook.id;
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookFiction.java:28]: wrong way of using super

File: src/main/java/Problem3/BookFiction.java [hunk 7]
```diff
@@ -24,6 +36,10 @@ public int getLateFeeInDollar() {
 
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
-        // homework
+        if (numOfDaysPastDue <=0){
+            return 0;
+        }else {
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookFiction.java:41]: "else" switch is not needed.

File: src/main/java/Problem3/BookRomance.java [hunk 8]
```diff
@@ -1,26 +1,42 @@
...
 
     private int lateFeePerDayInDollar = 4;
 
     public BookRomance(String title, String author) {
         // homework
         // tip: use the 'super' keyword
+        super();
+        super.title = title;
+        super.author = author;
+        super.id = UUID.randomUUID();
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookRomance.java:15]: here too. wrong way of using super and id generation

File: src/main/java/Problem3/BookRomance.java [hunk 9]
```diff
@@ -1,26 +1,42 @@
...
     }
 
     public BookRomance(BookRomance anotherBook) {
         // homework
         // tip: use the 'super' keyword
+        super();
+        super.title = anotherBook.title;
+        super.author = anotherBook.author;
+        super.id = anotherBook.id;
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookRomance.java:24]: here too. wrong way of using super.

File: src/main/java/Problem3/BookRomance.java [hunk 10]
```diff
@@ -1,26 +1,42 @@
...
+
         return lateFeePerDayInDollar;
     }
 
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
         // homework
+        if (numOfDaysPastDue <=0){
+            return 0;
+        }else {
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookRomance.java:38]: same here with "else"

File: src/main/java/Problem3/Movie.java [hunk 11]
```diff
@@ -8,16 +8,33 @@
...
     public boolean equals(Object obj) {
         // homework
+        if (obj == this){
+            return true;
+        }
+        if (!(obj instanceof Movie)){
+            return false;
+        }
+        Movie m1 = (Movie) obj;
+        return (this.id == m1.id && this.title == m1.title && this.rating == m1.rating);
```
Comment (pdgetrf) [inline src/main/java/Problem3/Movie.java:38]: wrong logic here too. only need to check id.

File: src/main/java/Problem3/MovieAction.java [hunk 12]
```diff
@@ -1,17 +1,28 @@
...
 
     private int lateFeePerDayInDollar = 5;
 
     public MovieAction(String rating, String title) {
         // homework
         // tip: use the 'super' keyword
+        super();
+        super.rating = rating;
+        super.title = title;
+        super.id = UUID.randomUUID();
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieAction.java:15]: here too. wrong way of using super and id generation

File: src/main/java/Problem3/MovieAction.java [hunk 13]
```diff
@@ -1,17 +1,28 @@
...
     }
 
     public MovieAction(MovieAction anotherMovie) {
         // homework
         // tip: use the 'super' keyword
+        super();
+        super.title = anotherMovie.title;
+        super.rating = anotherMovie.rating;
+        super.id = anotherMovie.id;
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieAction.java:24]: here too. wrong way of using super

File: src/main/java/Problem3/MovieAction.java [hunk 14]
```diff
@@ -22,5 +33,12 @@ public int getLateFeeInDollar() {
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
         // homework
+        if (numOfDaysPastDue >= 5){
+            return 2 * numOfDaysPastDue * getLateFeeInDollar();
+        }else if (numOfDaysPastDue <= 0) {
+            return 0;
+        }else{
+            return numOfDaysPastDue * getLateFeeInDollar();
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieAction.java:41]: duplicated logic here. 

suggesting:
```java
        int fee = lateFeePerDayInDollar * numOfDaysPastDue;
        return (numOfDaysPastDue >= 5) ? 2 * fee : fee;
```

File: src/main/java/Problem3/MovieComedy.java [hunk 15]
```diff
@@ -1,16 +1,26 @@
...
 public class MovieComedy extends Movie {
     private int lateFeePerDayInDollar = 3;
 
     public MovieComedy(String rating, String title) {
         // homework
         // tip: use the 'super' keyword
+        super();
+        super.rating = rating;
+        super.title = title;
+        super.id = UUID.randomUUID();
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieComedy.java:14]: here too. wrong way of using super and id generation

File: src/main/java/Problem3/MovieComedy.java [hunk 16]
```diff
@@ -20,6 +30,12 @@ public int getLateFeeInDollar() {
 
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
-        // homework
+
+        if (numOfDaysPastDue <=0){
+            return 0;
+        }else {
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieComedy.java:36]: same with else here

File: src/main/java/Problem3/MovieComedy.java [hunk 17]
```diff
@@ -1,16 +1,26 @@
...
+        super.id = UUID.randomUUID();
     }
 
     public MovieComedy(MovieComedy anotherMovie) {
         // homework
         // tip: use the 'super' keyword
+        super();
+        super.title = anotherMovie.title;
+        super.rating = anotherMovie.rating;
+        super.id = anotherMovie.id;
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieComedy.java:23]: here too. wrong way of using super

File: src/test/java/Problem2Test.java [hunk 18]
```diff
@@ -7,13 +7,21 @@
...
         };
 
         int[][] expects = {
-                {0, 1, 2, 4, 6}
-                // add more test cases
+                {0, 1, 2, 4, 6},
+                {0, 1, 2, 3, 4, 5},
+                {1, 2, 3, 4, 5},
+                {1, 3, 6, 7, 8, 9},
+                {2, 3, 5, 9}
```
Comment (pdgetrf) [inline src/test/java/Problem2Test.java:23]: great test. let's add more cases for the edge cases, such as null array and empty array.
